File: sandbox-worker/src/policy.ts
export const MAX_FILES = 1000;
export const MAX_EDITS = 100;
export const MAX_CHECKS = 10;
export const MAX_LOG_BYTES = 20_000;

const ALLOWED_EXECUTABLES = new Set([
  "python",
  "python3",
  "pytest",
  "ruff",
  "mypy",
  "pyright",
  "pip",
  "uv",
  "poetry",
  "node",
  "npm",
  "npx",
  "pnpm",
  "yarn",
  "bun",
  "go",
  "cargo",
  "bundle",
]);

const DENIED_PREFIXES = [
  ".git",
  ".github/workflows",
  ".circleci",
  ".buildkite",
  ".ssh",
  ".aws",
  ".config/gcloud",
];
const DENIED_FILE_NAMES = new Set([
  ".env",
  ".env.local",
  ".npmrc",
  ".pypirc",
  ".netrc",
  "id_rsa",
  "id_ed25519",
  "credentials",
  "credentials.json",
  "credentials.yml",
  "credentials.yaml",
  "service-account.json",
]);
const SECRET_PATTERNS = [
  /-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----/g,
  /\bgh[pousr]_[A-Za-z0-9_]{20,}\b/g,
  /\bgithub_pat_[A-Za-z0-9_]{20,}\b/g,
  /\bsk-[A-Za-z0-9_-]{20,}\b/g,
  /\b(?:api[_-]?key|secret|token|password)\s*[:=]\s*(['"]?)[^'"\s]{8,}\1/gi,
];

export type SandboxFile = { path: string; content_base64: string; sha256: string };
export type VerificationCommand = {
  id: string;
  kind: string;
  argv: string[];
  timeout_ms: number;
};
export type ExecutionRequest = {
  schema_version: "1.0";
  attempt_id: string;
  snapshot_digest: string;
  patch_digest: string;
  files: SandboxFile[];
  edits: SandboxFile[];
  checks: VerificationCommand[];
};

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export function normalizePath(value: string): string {
  if (!value || value.includes("\\") || value.includes("\0") || value.startsWith("/")) {
    throw new Error("invalid_path");
  }
  const parts = value.split("/");
  if (parts.some((part) => !part || part === "." || part === "..")) {
    throw new Error("invalid_path");
  }
  const normalized = parts.join("/");
  const lowered = normalized.toLowerCase();
  if (
    DENIED_PREFIXES.some(
      (prefix) => lowered === prefix || lowered.startsWith(`${prefix}/`),
    )
  ) {
    throw new Error("denied_path");
  }
  const name = parts.at(-1)?.toLowerCase() ?? "";
  if (
    DENIED_FILE_NAMES.has(name) ||
    name.startsWith(".env.") ||
    [".key", ".pem", ".p12", ".pfx"].some((suffix) => name.endsWith(suffix))
  ) {
    throw new Error("denied_path");
  }
  return normalized;
}

function parseFile(value: unknown): SandboxFile {
  if (!isRecord(value)) throw new Error("invalid_file");
  const path = value.path;
  const content = value.content_base64;
  const sha256 = value.sha256;
  if (
    typeof path !== "string" ||
    typeof content !== "string" ||
    typeof sha256 !== "string" ||
    !/^[a-f0-9]{64}$/.test(sha256) ||
    !/^[A-Za-z0-9+/]*={0,2}$/.test(content)
  ) {
    throw new Error("invalid_file");
  }
  return { path: normalizePath(path), content_base64: content, sha256 };
}

function parseCommand(value: unknown, maxTimeoutMs: number): VerificationCommand {
  if (!isRecord(value)) throw new Error("invalid_check");
  const { id, kind, argv, timeout_ms: timeoutMs } = value;
  if (
    typeof id !== "string" ||
    !/^[a-z0-9][a-z0-9_-]{0,99}$/.test(id) ||
    typeof kind !== "string" ||
    !Array.isArray(argv) ||
    argv.length < 1 ||
    argv.length > 32 ||
    typeof timeoutMs !== "number" ||
    !Number.isInteger(timeoutMs) ||
    timeoutMs < 100 ||
    timeoutMs > maxTimeoutMs
  ) {
    throw new Error("invalid_check");
  }
  if (!argv.every((argument) => typeof argument === "string")) {
    throw new Error("invalid_check");
  }
  const typedArgv = argv as string[];
  if (
    !ALLOWED_EXECUTABLES.has(typedArgv[0] ?? "") ||
    typedArgv.some(
      (argument) =>
        !argument || argument.length > 500 || /[\0\r\n]/.test(argument),
    )
  ) {
    throw new Error("invalid_check");
  }
  return { id, kind, argv: typedArgv, timeout_ms: timeoutMs };
}

export function parseExecutionRequest(value: unknown, maxTimeoutMs: number): ExecutionRequest {
  if (!isRecord(value)) throw new Error("invalid_request");
  const { schema_version: schemaVersion, attempt_id: attemptId } = value;
  const { snapshot_digest: snapshotDigest, patch_digest: patchDigest } = value;
  if (
    schemaVersion !== "1.0" ||
    typeof attemptId !== "string" ||
    !/^[A-Za-z0-9_-]{1,100}$/.test(attemptId) ||
    typeof snapshotDigest !== "string" ||
    !/^sha256:[a-f0-9]{64}$/.test(snapshotDigest) ||
    typeof patchDigest !== "string" ||
    !/^[a-f0-9]{64}$/.test(patchDigest) ||
    !Array.isArray(value.files) ||
    value.files.length > MAX_FILES ||
    !Array.isArray(value.edits) ||
    value.edits.length < 1 ||
    value.edits.length > MAX_EDITS ||
    !Array.isArray(value.checks) ||
    value.checks.length < 1 ||
    value.checks.length > MAX_CHECKS
  ) {
    throw new Error("invalid_request");
  }
  const files = value.files.map(parseFile);
  const edits = value.edits.map(parseFile);
  const checks = value.checks.map((check) => parseCommand(check, maxTimeoutMs));
  if (new Set(files.map((file) => file.path)).size !== files.length) {
    throw new Error("duplicate_file");
  }
  if (new Set(edits.map((file) => file.path)).size !== edits.length) {
    throw new Error("duplicate_edit");
  }
  if (new Set(checks.map((check) => check.id)).size !== checks.length) {
    throw new Error("duplicate_check");
  }
  return {
    schema_version: "1.0",
    attempt_id: attemptId,
    snapshot_digest: snapshotDigest,
    patch_digest: patchDigest,
    files,
    edits,
    checks,
  };
}

export function shellQuote(argument: string): string {
  return `'${argument.replaceAll("'", `'"'"'`)}'`;
}

export function redactLog(value: string): string {
  let redacted = value;
  for (const pattern of SECRET_PATTERNS) redacted = redacted.replace(pattern, "[REDACTED]");
  return redacted.slice(0, MAX_LOG_BYTES);
}

File: app/migration_generation/policy.py
"""Fail-closed path, patch, command, and secret policy."""

import base64
import hashlib
import json
import re
from pathlib import Path, PurePosixPath

from app.control_plane.models import FileChange, MigrationPlan, PatchEvidence
from app.repository_intelligence.workspace import _walk_entries

from .models import PatchProposal, SandboxExecutionRequest, SandboxFile

MAX_BUNDLE_FILES = 1000
MAX_BUNDLE_BYTES = 6_000_000
MAX_EDIT_BYTES = 2_000_000
DENIED_PATH_PREFIXES = (
    ".git",
    ".github/workflows",
    ".circleci",
    ".buildkite",
    ".ssh",
    ".aws",
    ".config/gcloud",
)
DENIED_FILE_NAMES = {
    ".env",
    ".env.local",
    ".npmrc",
    ".pypirc",
    "id_rsa",
    "id_ed25519",
    "credentials",
    "credentials.json",
    "credentials.yml",
    "credentials.yaml",
    "service-account.json",
    ".netrc",
}
SECRET_PATTERNS = (
    re.compile(r"-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----"),
    re.compile(r"\bgh[pousr]_[A-Za-z0-9_]{20,}\b"),
    re.compile(r"\bgithub_pat_[A-Za-z0-9_]{20,}\b"),
    re.compile(r"\bsk-[A-Za-z0-9_-]{20,}\b"),
    re.compile(
        r"(?i)\b(?:api[_-]?key|secret|token|password)\s*[:=]\s*"
        r"(['\"]?)[^'\"\s]{8,}\1"
    ),
)
ALLOWED_EXECUTABLES = {
    "python",
    "python3",
    "pytest",
    "ruff",
    "mypy",
    "pyright",
    "pip",
    "uv",
    "poetry",
    "node",
    "npm",
    "npx",
    "pnpm",
    "yarn",
    "bun",
    "go",
    "cargo",
    "bundle",
}


class PatchPolicyError(ValueError):
    code = "patch_policy_violation"


def normalize_repository_path(value: str) -> str:
    if not value or "\\" in value or "\x00" in value:
        raise PatchPolicyError("invalid repository path")
    path = PurePosixPath(value)
    if path.is_absolute() or any(part in {"", ".", ".."} for part in path.parts):
        raise PatchPolicyError("repository path escapes the workspace")
    normalized = path.as_posix()
    lowered = normalized.lower()
    if any(
        lowered == prefix or lowered.startswith(f"{prefix}/")
        for prefix in DENIED_PATH_PREFIXES
    ):
        raise PatchPolicyError("repository policy denies this path")
    if path.name.lower() in DENIED_FILE_NAMES or path.name.lower().startswith(".env."):
        raise PatchPolicyError("repository policy denies credential-bearing files")
    if path.suffix.lower() in {".key", ".pem", ".p12", ".pfx"}:
        raise PatchPolicyError("repository policy denies key material")
    return normalized


def contains_secret(value: str) -> bool:
    return any(pattern.search(value) for pattern in SECRET_PATTERNS)


def validate_argv(argv: list[str]) -> list[str]:
    if not argv or argv[0] not in ALLOWED_EXECUTABLES:
        raise PatchPolicyError("verification executable is not allowed")
    for argument in argv:
        if not argument or len(argument) > 500 or any(char in argument for char in "\x00\r\n"):
            raise PatchPolicyError("verification argument is invalid")
    return argv


def validate_patch(
    root: Path,
    plan: MigrationPlan,
    proposal: PatchProposal,
) -> tuple[bytes, PatchEvidence]:
    root = root.resolve()
    if contains_secret(json.dumps(proposal.model_dump(mode="json"), sort_keys=True)):
        raise PatchPolicyError("generated proposal contains credential-like material")
    plan_steps = {step.id: step for step in plan.steps}
    total_bytes = 0
    file_changes = []
    canonical_edits = []
    for edit in sorted(proposal.edits, key=lambda item: item.path):
        path = normalize_repository_path(edit.path)
        if unknown_steps := sorted(set(edit.plan_step_ids) - plan_steps.keys()):
            detail = ", ".join(unknown_steps)
            raise PatchPolicyError(f"edit references unknown plan steps: {detail}")
        expected_paths = {
            expected
            for step_id in edit.plan_step_ids
            for expected in plan_steps[step_id].expected_paths
        }
        if path not in expected_paths:
            raise PatchPolicyError("edit path is not approved by its plan step")
        encoded = edit.content.encode()
        total_bytes += len(encoded)
        if total_bytes > MAX_EDIT_BYTES:
            raise PatchPolicyError("generated patch exceeds byte limit")
        if contains_secret(edit.content):
            raise PatchPolicyError("generated patch contains credential-like material")
        destination = root / path
        if destination.is_symlink():
            raise PatchPolicyError("generated patch cannot modify a symbolic link")
        if destination.exists():
            if not destination.is_file() or not destination.resolve().is_relative_to(root):
                raise PatchPolicyError("generated patch target is outside the repository")
            existing = destination.read_bytes()
            expected = hashlib.sha256(existing).hexdigest()
            if edit.expected_sha256 != expected:
                raise PatchPolicyError("generated patch base digest is stale")
            change_type = "modified"
        else:
            if edit.expected_sha256 is not None:
                raise PatchPolicyError("new file cannot declare an existing-file digest")
            change_type = "added"
        canonical_edits.append(
            {
                "path": path,
                "expected_sha256": edit.expected_sha256,
                "content": edit.content,
                "plan_step_ids": sorted(edit.plan_step_ids),
            }
        )
        file_changes.append(
            FileChange(
                path=path,
                change_type=change_type,
                plan_step_ids=edit.plan_step_ids,
            )
        )
    for command in proposal.verification_commands:
        validate_argv(command.argv)
    artifact_bytes = json.dumps(
        {"schema_version": "1.0", "edits": canonical_edits},
        sort_keys=True,
        separators=(",", ":"),
    ).encode()
    digest = hashlib.sha256(artifact_bytes).hexdigest()
    return artifact_bytes, PatchEvidence(
        artifact_id=f"patch_{digest[:24]}",
        sha256=digest,
        summary=proposal.summary,
        files=file_changes,
    )


def build_sandbox_request(
    root: Path,
    attempt_id: str,
    snapshot_digest: str,
    patch: PatchEvidence,
    proposal: PatchProposal,
) -> SandboxExecutionRequest:
    root = root.resolve()
    files = []
    total_bytes = 0
    for relative, entry_type, value in sorted(_walk_entries(root), key=lambda item: item[0]):
        if entry_type != "file":
            continue
        try:
            path = normalize_repository_path(relative)
        except PatchPolicyError:
            continue
        content = value.read_bytes()
        try:
            if contains_secret(content.decode()):
                continue
        except UnicodeDecodeError:
            pass
        total_bytes += len(content)
        if len(files) >= MAX_BUNDLE_FILES or total_bytes > MAX_BUNDLE_BYTES:
            raise PatchPolicyError("repository exceeds sandbox upload limits")
        files.append(
            SandboxFile(
                path=path,
                content_base64=base64.b64encode(content).decode(),
                sha256=hashlib.sha256(content).hexdigest(),
            )
        )
    edits = [
        SandboxFile(
            path=normalize_repository_path(edit.path),
            content_base64=base64.b64encode(edit.content.encode()).decode(),
            sha256=hashlib.sha256(edit.content.encode()).hexdigest(),
        )
        for edit in proposal.edits
    ]
    return SandboxExecutionRequest(
        attempt_id=attempt_id,
        snapshot_digest=snapshot_digest,
        patch_digest=patch.sha256,
        files=files,
        edits=edits,
        checks=proposal.verification_commands,
    )
